Bug 5332: TCP_SWAPFAIL_MISS with UFS CF revalidation

eduard-bagdasaryan opened the pull request:
When collapsed_forwarding is enabled, collapsed revalidation requests
created UFS cache_dir entries with `ksRevalidation` key scope, which is
banned from caches. If Squid then decided to forward the being-cached
revalidation response to the client, `ksRevalidation` scope was cleared
in Store index, making the entry discoverable by regular transactions
(using `ksDefault` scope), even though on-disk entry bytes had a
`ksRevalidation` key. Attempts to load discovered entries from the cache
resulted in level-1 errors:

    ERROR: Failed to unpack Store entry metadata: swap meta MD5 mismatch
    exception location: SwapMetaIn.cc(89) CheckSwapMetaKey

2016 commit 1a210de4 placed `clearPublicKeyScope()` call inside
`clientReplyContext`, probably because it felt natural to clear
`ksRevalidation` scope in the same class that added it. That incorrect
placement resulted in entries with `ksRevalidation` keys getting cached.
Squid now clears `ksRevalidation` scope much earlier, in `makePublic()`,
before anything is written to Store, protecting caches from storing
`ksRevalidation` keys and narrowing revalidation-need window for
incoming requests.

This fix does not affect Squid caching decisions. For example,
`HttpStateData::reusableReply()` method has not changed and
`HttpStateData::haveParsedReplyHeaders()` still calls `makePublic()` for
responses that should be cached and `makePrivate()` for others.

This fix does not affect how Squid decides whether to give HTTP client
the old response, the current revalidation response, or restart from
scratch: `clientReplyContext::handleIMSReply()` conditions for calling
`sendClientOldEntry()`, `sendClientUpstreamResponse()`, or
`processMiss()` remain the same. Fixed code may `sendClientOldEntry()`
that got marked for release by adjusted `makePublic()`, but that marking
is not expected to affect what is being sent. Similarly, fixed code may
`sendClientUpstreamResponse()` using a public Store entry with
`ksDefault` scope instead of `ksRevalidate`, but that scope is not
expected to affect what is being sent.

Before this fix, some old entries were not purged because
`HttpStateData` called `makePublic()` (with the intent to cache the new
revalidation response and, hence, purge the old one), but `makePublic()`
did nothing for the already public `ksRevalidation` entries, and
`handleIMSReply()` went on `sendClientOldEntry()` path that did not call
`clearPublicKeyScope()`. Fixed code fulfills that `HttpStateData` intent
by storing cachable revalidation responses with `ksDefault` scope, which
is a positive change[^1].  We are not aware of such problems, but if
that intent is wrong in some cases, then those bugs should be fixed.
However, such bugs (if any) would also affect non-collapsed (i.e.
`ksDefault`) responses, and fixing all of that is out of scope here.

Also fixed `StoreEntry::mayStartSwapOut()` to not swapout private
entries.  The problem exists since 4310f8b. Storing private entries
violated Store metadata invariant: on startup, rock store produced
"Invalid entries" report.  UFS also stored private entries, but the
effect was hidden (except for disk IO overhead) because the just stored
disk entry was almost immediately removed by
`UFSSwapDir::evictCached()`.

[^1]: This change actually _restores_ the corresponding Squid behavior
prior to 2016 commit 1a210de4. We believe that old entry preservation
introduced by that commit in those cases was _accidental_. It may have
also been harmful because preserved old entry may keep triggering
(expensive) revalidation.

eduard-bagdasaryan committed:
Bug 5332: TCP_SWAPFAIL_MISS with UFS collapsed_forwarding

eduard-bagdasaryan committed:
Adjusted the previous solution

The new approach covers potential non-HTTP cases (although
processExpired() currently is only on HTTP path) and should be
more reliable (covering all possible cases, not only those
defined by preconditions in HttpStateData::haveParsedReplyHeaders()).

eduard-bagdasaryan committed:
Do not prohibit private entries in StoreEntry::swapOut()

Fixed a branch-added asertion.

rousskov commented on src/clients/Client.cc:
```diff
@@ -147,6 +147,9 @@ Client::setFinalReply(HttpReply *rep)
+        entry->clearPublicKeyScope();
+
     if (!EBIT_TEST(entry->flags, RELEASE_REQUEST) && blockCaching())
         entry->release();
```
Calling `release()` may make the entry private. To avoid expensive operations, all other factors being equal, it would be better to handle ksRevalidation scope (if it is still there) _after_ this call.

eduard-bagdasaryan committed:
Moved clearPublicKeyScope() into StoreEntry::startWriting()

We need to guarantee that clearPublicKeyScope() is called whenever we
write to cache (e.g., from replaceHttpReply()).

Also removed finalReplyIsFresh() precondition. The purpose was to cover
the 'sawDateGoBack' case in http.cc. However, if sawDateGoBack is true,
HttpStateData::reusableReply() returns 'reuseNot' and the entry
becomes private (and we do not need clearPublicKeyScope() for
private entries).

eduard-bagdasaryan commented on src/clients/Client.cc:
```diff
@@ -126,7 +126,7 @@ Client::setVirginReply(HttpReply *rep)
 
 HttpReply *
-Client::finalReply()
+Client::finalReply() const
```
> please move clearPublicKeyScope() code to StoreEntry::startWriting().

Moved locally (I'll push once GitHub fixes its connectivity problems). 

> check existing startWriting() and replaceHttpReply() callers

I could not find such suspicious cases. Some of the callers are definitely safe because they create StoreEntry  and call replaceHttpReply() in the same context.

rousskov commented on src/store.cc:
```diff
@@ -1726,6 +1735,8 @@ StoreEntry::startWriting()
     assert (isEmpty());
     assert(mem_obj);
 
+    clearPublicKeyScope();
```
Please check whether we can improve "no `ksRevalidation` keys in cache" guarantees further by moving this call to `makePublic(ksDefault)`. Currently, calling `makePublic(ksDefault)` has no effect on `ksRevalidation` entries. I wonder whether we can improve (and arguably even simplify) code by changing that design?

Current PR description contains a  TODO-marked paragraph that assumes this change was implemented without surprises (and explains why this change is a good idea). We will need to adjust that paragraph as needed (or completely replace it if you find reasons to reject this change request).

rousskov reviewed (changes requested):
Please also check the new PR title/description.

rousskov commented on src/clients/Client.cc:
> > 2014 commit [1a210de](https://github.com/measurement-factory/squid/commit/1a210de4cfe348c315f1334f4303ee73a45ef7cf) 
> 
> It is 2016 commit.

Fixed (in PR description).

> > original code before ... commit [1a210de](https://github.com/measurement-factory/squid/commit/1a210de4cfe348c315f1334f4303ee73a45ef7cf) purged the old entry as well
> 
> What line change of that commit removes the purging of the old entry?

The exact answer to your question feels meaningless/misleading to me because old entry preservation is caused by a _combination_ of changed _and unchanged_ code effects:

* Before that commit: revalidation entry is private; makePublic() call[^1] makes it public, releasing the old entry with the same key
* After that commit: revalidation entry is `ksRevalidate` public; makePublic() call[^1] does nothing, preserving the old entry with the same key

I have not tested the above theory. Also, it is possible that something else releases the old entry on the code path in question before commit 1a210de.

[^1]: On the code path we are discussing here, this `makePublic()` call is made by `cacheNegatively()`.

eduard-bagdasaryan committed:
Moved clearPublicKeyScope() into makePublic()

eduard-bagdasaryan committed:
Moved clearPublicKeyScope() further into setPublicKey()

eduard-bagdasaryan committed:
Refactored, moving makePrivate() to unsuccessful makePublic()

We need this change to handle unsuccessful ksRevalidation->ksDefault
switch, thus leaving no public entry behind on failure.

eduard-bagdasaryan committed:
Moved the swapout key check into StoreEntry::getSerialisedMetaData()

rousskov commented on src/clients/FtpGateway.cc:
```diff
@@ -2579,7 +2579,7 @@ Ftp::Gateway::haveParsedReplyHeaders()
     if (flags.authenticated || // authenticated requests can't be cached
             getCurrentOffset() ||
-            !e->makePublic()) {
+            !e->makePublic(false)) {
```
We have discussed that added `makePublic(false)` deserve analysis. I think this particular case matches the suggested "share if possible" default -- the caller does want to share this otherwise-cachable entry, even if it cannot be made public.

N.B. The effect of failed `makePublic()` is probably the same as of the `release()` call below. That existing `release()` call is OK and does not contradict `shareableWhenPrivate` "default" suggested for the `makePublic()` call above.

rousskov commented on src/store.cc:
```diff
@@ -699,7 +707,7 @@ StoreEntry::adjustVary()
         // then refactor to call storeCreatePureEntry() above.  Otherwise,
         // refactor to simply check whether `pe` is already public below.
-        if (!pe->makePublic()) {
+        if (!pe->makePublic(false)) {
```
We have discussed that added `makePublic(false)` deserve analysis. I think this particular case matches the suggested "share if possible" default -- there is no point in creating a Vary marker object if that object is not going to be shared beyond the current transaction, right?

rousskov commented on src/clients/FtpGateway.cc:
```diff
@@ -2579,7 +2579,7 @@ Ftp::Gateway::haveParsedReplyHeaders()
             getCurrentOffset() ||
-            !e->makePublic()) {
+            !e->makePublic(false)) {
         e->release();
```
I assume that in case of a failed makePublic() call failure, this `release()` call does nothing. Please correct me if that assumption is wrong.

P.S. I am aware that this `release()` call is needed for the other two `if` statement conditions.

eduard-bagdasaryan committed:
Fixed setPublicKey() to cover all 'public key scope changed' cases

eduard-bagdasaryan committed:
Fixed an assurance in getSerialisedMetaData()

eduard-bagdasaryan committed:
Revert "Fixed an assurance in getSerialisedMetaData()"

rousskov commented on src/clients/FtpGateway.cc:
```diff
@@ -2579,7 +2579,7 @@ Ftp::Gateway::haveParsedReplyHeaders()
-            !e->makePublic()) {
+            !e->makePublic(false)) {
         e->release();
     }
```
We have discussed that added `makePublic(false)` calls deserve analysis. I think this particular case matches the suggested default failure handling -- the caller does want to share this otherwise-cachable entry among existing readers, even if it cannot be made `ksDefault` public. That also implies that this caller does not want to call `release(void)` if `makePublic()` fails because that call would make the entry not just private but non-sharable.

On the other hand, the `release()` call is needed for the other two `if` statement conditions and, hence, cannot be simply removed.

I suggest the following refactoring:

```suggestion
if (flags.authenticated || getCurrentOffset()) {
    e->release();
    return;
}
e->makePublic();
```

I do not think we need existing C++ comments -- this code is self-documenting enough IMO.

rousskov reviewed (changes requested):
I hope this partial review allows us to progress further.

rousskov commented on src/store.cc:
```diff
@@ -1749,6 +1752,8 @@ StoreEntry::startWriting()
 StoreEntry::getSerialisedMetaData(size_t &length) const
 {
+    // allow only private or public default entries
+    Assure(!publicKey() || !publicDefaultKeyCmp());
```
I see no reason to write private entries to disk (when disk cache is enabled). We should be using local working memory for them (that is always there). However, the triggered assertion does indicate a problem or bug. More research is needed to see whether we can avoid starting swapout of private entries.

eduard-bagdasaryan replied:
This particular problem with `truncacate-responses.js` happens only with a body size less than 4K (`SM_PAGE_SIZE`) because `StoreEntry::swapOut()` postpones disk writing due to the `swapout_maxsize < SM_PAGE_SIZE` check. When the entry becomes private (after `httpTimeout()` and `lengthWentBad()`), `mayStartSwapOut()` proceeds further, because `mayStartSwapOut()` is  "already allowed" and `SM_PAGE_SIZE` is not checked (store_status is `STORE_OK` now).

In general, private entries (that are already private at the first `swapOut()` call) are not written to disk because `StoreEntry::mayStartSwapOut()` returns false after `StoreEntry::checkCachable()` checks that `RELEASE_REQUEST` is set.

rousskov replied:
Thank you for that archeological discovery! I agree that we are dealing with `publish_mayStartSwapOut_delete_startSwapOut` case here; there are even Jira [comments](https://measurementfactory.atlassian.net/browse/SQUID-175?focusedCommentId=24107) that specifically describe significant delays and release-triggering events that may happen between `mayStartSwapOut()` and `startSwapOut()`. Your interpretation regarding `markedForDeletionAndAbandoned()` optimization also sounds plausible to me.

The table explains that we

eduard-bagdasaryan committed:
Refactored makePublic() into makePublic() and makePublicWith()

Depending on the caller type/purpose one of the new methods is used.

eduard-bagdasaryan commented on src/store/Controller.cc:
```diff
@@ -730,7 +730,7 @@ Store::Controller::allowCollapsing(StoreEntry *e, const RequestFlags &reqFlags,
     // set the flag now so that it gets copied into the Transients entry
     e->setCollapsingRequirement(true);
-    if (e->makePublic(keyScope)) { // this is needed for both local and SMP collapsing
+    if (e->makePublic(false, keyScope)) { // this is needed for both local and SMP collapsing
```
Done.

eduard-bagdasaryan commented on src/Store.h:
```diff
@@ -96,17 +96,18 @@ class StoreEntry : public hash_link, public Packable
-    void clearPublicKeyScope();
+
+    /// \sa makePublic()
+    bool setPublicKey(bool shareable, KeyScope keyScope = ksDefault);
```
Ok.

>  it is possible to implement makePublic(void) using makePublicWIth()

If I got your intention correctly, makePublic() simply calls makePublicWith(ksDefault). None of the methods should makePrivate(false). If a caller needs makePrivate(false), it calls it after makePublicWith() failure. Currently there are no such cases.

eduard-bagdasaryan committed:
Undone an out-of-scope change

eduard-bagdasaryan reviewed (commented):
Applied refactoring suggestions.

eduard-bagdasaryan committed:
Fixed a mistake in StoreEntry::makePublicWith()

rousskov commented on src/tests/stub_store.cc:
```diff
@@ -36,7 +36,8 @@ void StoreEntry::replaceHttpReply(const HttpReplyPointer &, bool) STUB
 void StoreEntry::abort() STUB
-bool StoreEntry::makePublic(const KeyScope) STUB
+void StoreEntry::makePublic() STUB
+bool StoreEntry::makePublicWith(const KeyScope) STUB
```
```suggestion
bool StoreEntry::makePublicWith(KeyScope) STUB
```

rousskov commented on src/store.cc:
```diff
@@ -163,11 +163,25 @@ StoreEntry::operator delete (void *address)
     /* This object can be cached for a long time */
-    return !EBIT_TEST(flags, RELEASE_REQUEST) && setPublicKey(scope);
+    if (EBIT_TEST(flags, RELEASE_REQUEST))
+        return false;
```
Missing `makePrivate(true)` AFAICT -- the entry may not have been marked as valid for sharing before... And if we cannot call `makePrivate(true)` here, we need to change this method description as well!

To reduce code duplication, this method implementation may need to be refactored as 

```C++
if (!EBIT_TEST(flags, RELEASE_REQUEST) && setPublicKey(scope))
    return true;

makePrivate(true); // may already be private and shareableWhenPrivate
return false;
```

eduard-bagdasaryan committed:
Made StoreEntry::cacheNegatively() unconditional and other polishing

eduard-bagdasaryan reviewed (commented):
Adjusted (63d3baf).

eduard-bagdasaryan committed:
Fixed tests

eduard-bagdasaryan committed:
Added a comment

eduard-bagdasaryan committed:
Do not swapout private entries

eduard-bagdasaryan committed:
Simplified, inlining Controller::privateOrMarkedForDeletion()

and separated debugging.

rousskov committed:
fixup: Documented a concern for future SMP collapsed revalidation work

rousskov committed:
fixup: Describe branch-modified cacheNegatively()

... instead of dragging its poorly formatted description TODO across
this branch changes.

TODO: Make negativeCache() private. It is already described in a .cc
file, so we just need to move its declaration to the private section.

rousskov committed:
fixup: Made branch-added publicDefaultKeyCmp() private

This API is fairly complex/risky. Others should not be using it.

rousskov committed:
fixup: Significantly simplified publicDefaultKeyCmp() API

In the earlier variations of this refactoring, I incorrectly interpreted
publicDefaultKeyCmp() API. Twice!
